modules/sitemaps/sitemap-finder.php

```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * The functions in this class provide an API for handling
+ * sitemap related URIs.
+ *
+ * @package Jetpack
+ * @since 4.6.0
+ * @author Automattic
+ */
+
+/**
+ * The Jetpack_Sitemap_Finder object deals with constructing
+ * sitemap URIs.
+ *
+ * @since 4.6.0
+ */
+class Jetpack_Sitemap_Finder {
+
+	/**
+	 * Jetpack_Sitemap_Finder constructor.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 */
+	public function __construct() {
+		return;
+	}
+
+	/**
+	 * Construct the complete URL of a sitemap file. Depends on
+	 * permalink settings.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @param string $filename The filename of the sitemap.
+	 *
+	 * @return string Complete URI of the given sitemap file.
+	 */
+	public function construct_sitemap_url( $filename ) {
+		global $wp_rewrite;
+
+		$location = Jetpack_Options::get_option_and_ensure_autoload(
+			'jetpack_sitemap_location',
+			''
+		);
+
+		if ( $wp_rewrite->using_index_permalinks() ) {
+			return home_url( '/index.php' . $location . '/' . $filename );
+		} elseif ( $wp_rewrite->using_permalinks() ) {
+			return home_url( $location . '/' . $filename );
+		} else {
+			return home_url( $location . '/?jetpack-sitemap=' . $filename );
+		}
+	}
+
+	/**
+	 * Path and query prefix of sitemap files. Depends on permalink
+	 * settings.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @return string The path+query prefix.
+	 */
+	public function the_jetpack_sitemap_path_and_query_prefix() {
+		global $wp_rewrite;
+
+		// Get path fragment from home_url().
+		$home = wp_parse_url( home_url() );
+		if ( isset( $home['path'] ) ) {
+			$home_path = $home['path'];
+		} else {
+			$home_path = '';
+		}
+
+		// Get additional path fragment from filter.
+		$location = Jetpack_Options::get_option_and_ensure_autoload(
+			'jetpack_sitemap_location',
+			''
+		);
+
+		if ( $wp_rewrite->using_index_permalinks() ) {
+			return $home_path . '/index.php' . $location . '/';
+		} elseif ( $wp_rewrite->using_permalinks() ) {
+			return $home_path . $location . '/';
+		} else {
+			return $home_path . $location . '/?jetpack-sitemap=';
+		}
+	}
+
+	/**
+	 * Examine the path of the request URI looking for a sitemap request.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @return array @args {
+	 *   @type string $sitemap_path The path where sitemaps are served.
+	 *   @type string $raw_uri The raw request URI.
+	 *   @type string $sitemap_uri The recognized sitemap path fragment (or null).
+	 *   @type string $sitemap_name The recognized sitemap name (or null).
+	 * }
+	 */
+	public function recognize_sitemap_uri() {
```

Comment:
The plugin works by looking at the incoming `REQUEST_URI`, and then if that URI corresponds to a sitemap file (of which there are several types) it serves the file from the database and dies.

This is the function that takes the `REQUEST_URI`, determines whether it corresponds to a sitemap, and if so, determines *which* sitemap file it corresponds to. It recognizes sitemap URIs.

So for instance if the request URI is `example.com/sitemap.xml` this returns an array whose relevant parts are `sitemap_uri => "sitemap.xml"` and `sitemap_name => "sitemap"`. The `Jetpack_Sitemap_Manager` object then regex matches on `sitemap_uri` to decide what type of file to serve, and `sitemap_name` to decide which file to serve. (Keeping track of both names and types seems redundant, but is needed elsewhere.)

but yeah this could use some cleaning up.